src/p_sparse_matrix.jl

```diff
@@ -365,17 +360,11 @@ end
 
 function Base.similar(a::PSparseMatrix,::Type{T},inds::Tuple{<:PRange,<:PRange}) where T
     rows,cols = inds
-    matrix_partition = map(partition(a),partition(rows),partition(cols)) do values, row_indices, col_indices
+    row_partition = partition(rows)
+    col_partition = partition(cols)
+    matrix_partition = map(partition(a),row_partition,col_partition) do values, row_indices, col_indices
         allocate_local_values(values,T,row_indices,col_indices)
     end
-    PSparseMatrix(values,row_partition,col_partition)
-end
-
-function Base.similar(::Type{<:PSparseMatrix{V}},inds::Tuple{<:PRange,<:PRange}) where V
```

Comment:
I think we need to keep this method. Any good reason for deleting it?

Reply:
Yes. At least I do not see how the local portion of the returned instance can be fully specified (i.e, nrows, ncols, and sparsity pattern) solely with the type `V`. We might define a local portion with zero nonzero elements, but I do not find much sense to it. Do you agree?

Reply:
I note this is not the case with `::Type{<:PVector{V}}`, for which this method is properly defined from `V`.

Reply:
I see, I would follow the same approach as for the standard SparseMatrixCSC matrix.  I guess the problem there is conceptually the same, right?

Reply:
Yes, I agree it is the same problem.

Unfortunately, the call to `similar(SparseMatrixCSC{T1,T2}, m, n)` returns an error.

Reply:
If the error is fixed in SparseArrays, would our method work? If yes, I would keep our method.

Reply:
Another point to consider: SparseMatrixCSC is not the only type possible to represent the local portions of the partitioned matrix. Other container types might implement the similar method for these types.

Reply:
For the records, the error with SparseMatrixCSC is:

```
ERROR: MethodError: no method matching SparseArrays.SparseMatrixCSC{Float64, Int64}(::UndefInitializer, ::Tuple{Int64, Int64})

Closest candidates are:
  SparseArrays.SparseMatrixCSC{Tv, Ti}(::LinearAlgebra.UniformScaling, ::Tuple{Int64, Int64}) where {Tv, Ti}
   @ SparseArrays ~/software_installers/julia-1.9.1/share/julia/stdlib/v1.9/SparseArrays/src/sparsematrix.jl:2016
  SparseArrays.SparseMatrixCSC{Tv, Ti}(::UndefInitializer, ::Integer, ::Integer) where {Tv, Ti}
   @ SparseArrays ~/software_installers/julia-1.9.1/share/julia/stdlib/v1.9/SparseArrays/src/sparsematrix.jl:55

Stacktrace:
  [1] similar(#unused#::Type{SparseArrays.SparseMatrixCSC{Float64, Int64}}, dims::Tuple{Int64, Int64})
    @ Base ./abstractarray.jl:882
  [2] similar(::Type{SparseArrays.SparseMatrixCSC{Float64, Int64}}, ::Int64, ::Int64)
    @ Base ./abstractarray.jl:880
  [3] allocate_local_values(#unused#::Type{SparseArrays.SparseMatrixCSC{Float64, Int64}}, indices_rows::PartitionedArrays.LocalIndicesWithConstantBlockSize{1}, indices_cols::PartitionedArrays.LocalIndicesWithConstantBlockSize{1})
    @ PartitionedArrays ~/git-repos/PartitionedArrays.jl/src/p_sparse_matrix.jl:15
  [4] (::PartitionedArrays.var"#474#475"{SparseArrays.SparseMatrixCSC{Float64, Int64}})(row_indices::PartitionedArrays.LocalIndicesWithConstantBlockSize{1}, col_indices::PartitionedArrays.LocalIndicesWithConstantBlockSize{1})
    @ PartitionedArrays ~/git-repos/PartitionedArrays.jl/src/p_sparse_matrix.jl:382
  [5] (::Base.var"#4#5"{PartitionedArrays.var"#474#475"{SparseArrays.SparseMatrixCSC{Float64, Int64}}})(a::Tuple{PartitionedArrays.LocalIndicesWithConstantBlockSize{1}, PartitionedArrays.LocalIndicesWithConstantBlockSize{1}})
    @ Base ./generator.jl:36
  [6] iterate
    @ ./generator.jl:47 [inlined]
  [7] collect
    @ ./array.jl:782 [inlined]
  [8] map
    @ ./abstractarray.jl:3383 [inlined]
  [9] map
    @ ~/git-repos/PartitionedArrays.jl/src/debug_array.jl:105 [inlined]
 [10] similar(#unused#::Type{PartitionedArrays.PSparseMatrix{SparseArrays.SparseMatrixCSC{Float64, Int64}, PartitionedArrays.DebugArray{SparseArrays.SparseMatrixCSC{Float64, Int64}, 1}, PartitionedArrays.DebugArray{PartitionedArrays.LocalIndicesWithConstantBlockSize{1}, 1}, PartitionedArrays.DebugArray{PartitionedArrays.LocalIndicesWithConstantBlockSize{1}, 1}, PartitionedArrays.DebugArray{PartitionedArrays.SparseMatrixAssemblyCache, 1}, Float64}}, inds::Tuple{PartitionedArrays.PRange{PartitionedArrays.DebugArray{PartitionedArrays.LocalIndicesWithConstantBlockSize{1}, 1}}, PartitionedArrays.PRange{PartitionedArrays.DebugArray{PartitionedArrays.LocalIndicesWithConstantBlockSize{1}, 1}}})
    @ PartitionedArrays ~/git-repos/PartitionedArrays.jl/src/p_sparse_matrix.jl:381
 [11] p_sparse_matrix_tests(distribute::Type{PartitionedArrays.DebugArray})
    @ Main.DebugArrayPSparseMatrixTests ~/git-repos/PartitionedArrays.jl/test/p_sparse_matrix_tests.jl:60
 [12] with_debug(f::typeof(Main.DebugArrayPSparseMatrixTests.p_sparse_matrix_tests))
    @ PartitionedArrays ~/git-repos/PartitionedArrays.jl/src/debug_array.jl:8
 [13] top-level scope
    @ ~/git-repos/PartitionedArrays.jl/test/debug_array/p_sparse_matrix_tests.jl:7
```

Reply:
> If the error is fixed in SparseArrays, would our method work? If yes, I would keep our method.

Yes, it would work, but what's the point of creating a sparse matrix with an arbitrary sparsity pattern?

Reply:
>  Another point to consider: SparseMatrixCSC is not the only type possible to represent the local portions of the partitioned matrix. Other container types might implement the similar method for these types.

Ok, that is a fair point. We can, e.g., have dense matrices as the local portion of PSparseMatrix. For dense matrices, `similar(V,m,n)` is well defined. I will keep the method. Reverting changes.

Reply:
> Ok, that is a fair point. We can, e.g., have dense matrices as the local portion of PSparseMatrix. For dense matrices, similar(V,m,n) is well defined. I will keep the method. Reverting changes.

I already did this (i.e., reverting). I also tried to add a test for the controversial variant of `similar()` with a dense matrix as local portion. However, it seems that the library is not actually prepared to handle local portions which are dense matrices (I think the previous version of PartitionedArrays could handle this, although I might be wrong). In particular, the APIs which are leveraged by `p_sparse_matrix_cache_impl` are not supported by dense matrices. I implemented `nziterator` for `AbstractMatrix`:

```
+nziterator(a::AbstractMatrix) = NZIteratorAbstractMatrix(a)
+
+struct NZIteratorAbstractMatrix{A}
+    array::A
+end
+
+Base.length(a::NZIteratorAbstractMatrix) = prod(size(a.matrix))
+Base.eltype(::Type{<:NZIteratorAbstractMatrix{A}}) where A = Tuple{Int,Int,eltype(A)}
+Base.eltype(::T) where T <: NZIteratorAbstractMatrix = eltype(T)
+Base.IteratorSize(::Type{<:NZIteratorAbstractMatrix}) = Base.HasLength()
+Base.IteratorEltype(::Type{<:NZIteratorAbstractMatrix}) = Base.HasEltype()
+
+@inline function Base.iterate(a::NZIteratorAbstractMatrix)
+    if (length(a.array)==0) 
+        return nothing 
+    end 
+    cids=CartesianIndices(size(a.array))
+    cid,state_cids_it=iterate(cids)
+    (Tuple(cid)...,a.array[cid]), (cids,state_cids_it)
+end
+
+@inline function Base.iterate(a::NZIteratorAbstractMatrix,state)
+    (cids,state_cids_it) = state
+    b=iterate(cids,state_cids_it)
+    if (b==nothing)
+      return nothing
+    end
+    cid,state_cids_it=b
+    (Tuple(cid)...,a.array[cid]), (cids,state_cids_it)
+end
+
```

However, then I hit `SubSparseMatrix` ... and I gave up. Just reporting here for future reference.

Long history short, I cannot come up with a call to the controversial `similar()` function which does not break.

Reply:
@fverdugo  ... any thoughts on this?

Reply:
> @fverdugo ... any thoughts on this?

I would say this is not mission critical at this moment. 

> (I think the previous version of PartitionedArrays could handle this, although I might be wrong

The nziterator, SubSparseMatrix and et al are (almost) verbatim from version v0.2.